sds/src/match_validation/aws_validator.rs

```diff
@@ -1,22 +1,145 @@
+    async fn validate(
+        &self,
+        matches: &mut Vec<RuleMatch>,
+        scanner_rules: &Vec<Box<dyn CompiledRuleDyn>>,
+    ) {
+        // Let's regroup matches per type
+        let mut aws_id_matches_idx = vec![];
+        let mut aws_secret_matches_idx = vec![];
+
+        for (idx, m) in matches.iter().enumerate() {
+            let rule: &Box<dyn CompiledRuleDyn> = &scanner_rules[m.rule_index];
+            match rule.get_match_validation_type() {
+                Some(MatchValidationType::Aws(AwsType::AwsId)) => {
+                    aws_id_matches_idx.push(idx);
+                }
+                Some(MatchValidationType::Aws(AwsType::AwsSecret(_))) => {
+                    aws_secret_matches_idx.push(idx);
+                }
+                _ => {}
+            }
+        }
+
+        let mut match_status_per_pairs_of_matches_idx: AHashMap<(usize, usize), MatchStatus> =
+            AHashMap::new();
+        for id_index in 0..aws_id_matches_idx.len() {
+            for secret_index in 0..aws_secret_matches_idx.len() {
+                match_status_per_pairs_of_matches_idx.insert(
+                    (
+                        aws_id_matches_idx[id_index],
+                        aws_secret_matches_idx[secret_index],
+                    ),
+                    MatchStatus::NotChecked,
+                );
+            }
+        }
+
+        let client = Client::new();
+        // Let's try all combination of aws_id and aws_secret
+        let futures = match_status_per_pairs_of_matches_idx.iter_mut().map(
+            |((id_index, secret_index), match_status)| {
+                let client = client.clone();
+                let match_id = &matches[*id_index];
+                let match_secret = &matches[*secret_index];
+                async move {
+                    // Let's reqwest the HTTP API endpoint to validate the matches
+                    let mut datetime = chrono::Utc::now();
+                    if !self.config.forced_datetime_utc.is_none() {
+                        datetime = self.config.forced_datetime_utc.unwrap()
+                    }
+                    let (body, headers) = generate_aws_headers_and_body(
+                        &datetime,
+                        &self.config.aws_sts_endpoint,
+                        &match_id.match_value.as_ref().unwrap(),
+                        &match_secret.match_value.as_ref().unwrap(),
+                    );
+                    let res = client
+                        .post(self.config.aws_sts_endpoint.as_str())
+                        .headers(headers)
+                        .body(body)
+                        .timeout(self.config.timeout)
+                        .send()
+                        .await;
+
+                    match res {
+                        Ok(val) => {
+                            // If status is 200-299, then it's valid we can safely update the match status
```

Comment:
It seems like you are making this more generic later (validators can pick ranges that are considered "success"?) but just wanted to make sure you knew that the `StatusCode` type has methods to do this for you if it helps at all: https://docs.rs/reqwest/0.12.7/reqwest/struct.StatusCode.html#method.is_success

Reply:
True I will use it for aws

Reply:
done